SerialPrograms/Source/Integrations/DppIntegration/DppClient.cpp

```diff
@@ -0,0 +1,155 @@
+				const JsonObject& json_obj,
+                const std::string& message,
+				std::shared_ptr<PendingFileSend> file) {
+                if (m_is_connected) {
+                    embed embed;
+                    {
+                        embed.set_title(*json_obj.get_string("title"));
+                        embed.set_color((int)((uint32_t)color & 0xffffff));
+
+                        auto fields = json_obj.get_array("fields");
+                        for (auto& field : *fields) {
+                            auto obj = field.get_object();
+                            embed.add_field(*obj->get_string("name"), *obj->get_string("value"));
+                        }
+                    }
+
+                    std::set<std::string> tag_set;
+                    for (const std::string& tag : tags) {
+                        tag_set.insert(to_lower(tag));
+                    }
+
+                    DiscordSettingsOption& settings = GlobalSettings::instance().DISCORD;
+                    const DiscordIntegrationTable& channels = settings.integration.channels;
+                    std::vector<std::string> channel_vector;
+                    std::vector<std::string> message_vector;
+
+                    std::vector<std::unique_ptr<DiscordIntegrationChannel>> table = channels.copy_snapshot();
+                    for (size_t i = 0; i < table.size(); i++) {
+                        const Integration::DiscordIntegrationChannel& channel = *table[i];
+                        if (((std::string)channel.tags_text).empty() || !channel.enabled) {
+                            continue;
+                        }
+
+                        bool send = false;
+                        for (const std::string& tag : EventNotificationOption::parse_tags(channel.tags_text)) {
+                            auto iter = tag_set.find(to_lower(tag));
+                            if (iter != tag_set.end()) {
+                                channel_vector.emplace_back(channel.channel_id);
+                                send = true;
+                                break;
+                            }
+                        }
+
+                        if (!send) {
+                            continue;
+                        }
+
+                        if (std::atoll(((std::string)settings.message.user_id).c_str()) == 0) {
+                            should_ping = false;
+                        }
+
+                        std::string msg = "";
+                        if (should_ping && channel.ping) {
+                            msg += "<@" + (std::string)settings.message.user_id + ">";
+                        }
+
+                        const std::string& discord_message = settings.message.message;
+                        if (!discord_message.empty()) {
+                            if (!msg.empty()) {
+                                msg += " ";
+                            }
+
+                            for (char ch : discord_message) {
```

Comment:
doing this isn't great for performance reason. You are resizing the string a lot.

discord_message.erase(std::remove(discord_message.begin(), discord_message.end(), "@"), discord_message.end());
msg += discord_message;

plays better with memory allocation because you resize only once

Reply:
I honestly just copied the embed building from Sleepy, not even sure if it was mine or Kuro's 😆